plutus-playground/plutus-playground-lib/src/Playground/Interpreter/Util.hs

```diff
@@ -0,0 +1,236 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module Playground.Interpreter.Util where
+
+import           Control.Monad.Error.Class  (MonadError, throwError)
+import           Data.Aeson                 (FromJSON)
+import qualified Data.Aeson                 as JSON
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import qualified Data.Map                   as Map
+import qualified Data.Set                   as Set
+import qualified Data.Typeable              as T
+import Control.Lens (view)
+import Data.Monoid (Sum(Sum), getSum)
+import           Playground.API             (PlaygroundError (OtherError))
+import           Wallet.Emulator.Types      (walletStates, ownFunds, EmulatorEvent, EmulatedWalletApi, EmulatorState (_chainNewestFirst, _emulatorLog), Trace,
+                                             Wallet (Wallet), runTraceTxPool)
+import           Wallet.Generators          (GeneratorModel (GeneratorModel))
+import qualified Wallet.Generators          as Gen
+import           Wallet.UTXO                (Blockchain, PubKey (PubKey), Tx, Value (Value))
+
+-- | Unfortunately any uncaught errors in the interpreter kill the thread that is running it rather than returning the error. This means we need to handle all expected errors in the expression we are interpreting. This gets a little tricky because we have to decode JSON inside the interpreter (since we don't have access to it's type outside) so we need to wrap the @apply functions up in something that can throw errors.
+runTrace ::
+     [(Wallet, Int)]
+  -> [Either PlaygroundError (Trace EmulatedWalletApi [Tx])]
+  -> Either PlaygroundError (Blockchain, [EmulatorEvent], [(Wallet, Value)])
+runTrace wallets actions =
+  let walletToBalance (Wallet i, v) = (PubKey i, Value v)
+      initialBalance = Map.fromList $ fmap walletToBalance wallets
+      pubKeys = Set.fromList $ fmap (\(Wallet i, _) -> PubKey i) wallets
+      eActions = sequence actions
+   in case eActions of
+        Left e -> Left e
+        Right actions' ->
+          let action = sequence actions'
```

Comment:
Why not

```haskell
notifyAll = processPending >>= walletsNotifyBlock (Wallet <$> [1.. length wallets])
action = notifyAll >> sequence actions'
```
?
that way you always get an initial block with the initial transaction, even if the user-submitted list of actions is empty. 

You could even do

```
action = notifyAll >> sequence (fmap (\a -> a >> notifyAll) actions'
```

Then you don't need `runWalletActionAndProcessPending` anymore (which is called in a big concatenated string)´. That would make it more transparent.

Reply:
done first suggestion but didn't do second as it would make the string building function a bit awkward with little benefit IMO